sympy/matrices/random.py

```diff
@@ -0,0 +1,2378 @@
+from ..core import I as _i
+from ..core.mul import Mul as _multiply
+from ..core.random import sample as _smple
+from ..core.sympify import _sympify
+from ..functions import sqrt as _sqrt, re as _re, im as _im, \
+    transpose as _t, adjoint as _c
+from ..tensor import shape as _shape
+from .dense import eye as _eye
+from .inverse import _inv
+
+__all__ = 'projection', 'jordan', 'transposition', \
+          'permutation', 'elementary', 'rotation', 'reflection', \
+          'diagonal_normal', 'jordan_normal', 'isometry_normal', \
+          'triangular', 'invertible', 'singular', \
+          'idempotent', 'nilpotent', 'diagonalizable', 'trigonalizable', \
+          'orthogonal', 'unitary', 'normal', \
+          'symmetric', 'hermite', 'square', \
+          'regular_to_singular', 'complex_to_real'
+
+
+_EPS = 1e-15
+
+
+# === random number generator functions ===
+
+
+def _sample(scalars, k=None):
+    _k = k or 1
+    if hasattr(scalars, 'sample'):
+        smpl = scalars.sample(_k)
+    else:
+        smpl = _smple(scalars, _k)
+    return smpl if k else smpl[0]
+
+
+# === helper ===
+
+
+def _cs(scalar, real=True):
+    if isinstance(scalar, (tuple, list)) and len(scalar) == 2:
+        c, s = scalar
+    elif _is_complex(scalar):
+        c, s = _re(scalar), _im(scalar)
+    else:
+        c, s = scalar, _sample((-1, 1)) * _sqrt(1 - scalar ** 2)
+    abs_cs = abs(c ** 2 + s ** 2 - 1)
+    if isinstance(scalar, (int, float, complex)):
+        if abs(scalar) > 1 or abs_cs > _EPS:
```

Comment:
I think that this can just be `if (c**2 + s**2 - 1).is_zero is False:` or using the internal `is_ne` function `if is_ne(c**2 + s**2, 1):`

Reply:
simplified this part by removing the `abs_cs > _EPS` condition. But something similar happens in `_is_abs_one`. Currently I use buildins `isclose` as a strikt compare fails for
```
>>> z = complex(1, 2)
>>> abs(z/abs(z)) == 1
False
```

Reply:
If there is any way to improve this, I would be happy to adopt it.

Reply:
same for `_is_complex`

Reply:
We have:
```
def _is_abs_one(x):
    return isinstance(abs(x), (int, float, Number)) and isclose(abs(x), 1)
```
This is basically just `x in [-1, 1]` if I understand correctly. Why do we need the `isclose`? There shouldn't be any rounding error.

Reply:
Oh, I see the idea is that if `x` is complex then `abs(x)` could be float.

In general it is better to work with sympified objects rather than having `int`, `float` or `complex`. The idea is that when receiving input from the user you call `_sympify` on it and then you know everywhere internally that you have symbolic expressions.

Reply:
With exact symbolic expressions there shouldn't be any rounding error so no need for `isclose` etc.

Reply:
I'm not sure what the purpose of `_is_complex` is supposed to be.

Reply:
Replaced `_abs_one(x)` by `abs(x) == 1` to avoid `isclose`. Note that as `complex` has numerical issues, will not work some cases. Even not for
```
>>> z = sympify(complex(1, 2))
>>> abs(z/abs(z)) == 1
False
```

> I'm not sure what the purpose of `_is_complex` is supposed to be.

replaced `_is_complex` by `_is_real` in order to check if the values have no imaginary part.